### DISCUSSION TITLE
How do you impersonate a user when testing? (Mock a session)

### DISCUSSION THREAD

**Discussion by vincentrolfs** [blitz-js/blitz#2653] [Q&A]
I want to write an integration test where I create a user and then call a query while acting as this user. Essentially I need to mock the session for that user. What is the standard way to do this in blitz?

I could not find any information about this in the [testing section](https://blitzjs.com/docs/testing) or the [session section](https://blitzjs.com/docs/session-management).

Thanks a lot!

EDIT: More specifically, *I need to test the authorization capabilities of the app*, so mocking the session in the following is not sufficient:
```
    const mockCtx: any = {
      session: {
        $create: jest.fn,
        $authorize: () => {},
        userId: user.id,
      },
    }
```

I want to say "pretend I am logged in as this user and I call that query, what happens?"

**Comment by Talor-A** [comment]
in `test/utils.tsx`:

```tsx
export function getTestSession({
  user,
  orgId,
  ...rest
}: Partial<Ctx> & { user: User; orgId?: number }): Ctx {
  return {
    session: {
      $authorize: jest.fn(),
      $isAuthorized: jest.fn(() => true),
      userId: user?.id,
      roles: ["USER"],
      $getPrivateData: jest.fn(),
      $create: jest.fn(),
      $handle: null,
      $publicData: {
        roles: ["USER"],
        userId: user.id,
        orgId,
      },
      $revoke: jest.fn(),
      $revokeAll: jest.fn(),
      $setPrivateData: jest.fn(),
      $setPublicData: jest.fn(),
      orgId,

      ...rest.session,
    },
  }
}
```

and then use it like below. this is test assertions from my `inviteUser` mutation that adds users to the organization. it relies on `orgId` being set in the inviter's session, to know which organization to add to.

```tsx
expect(async () =>
  inviteUser(
    {
      asRole: "USER",
      invitedUserEmail: "inviteduseremail@test.com",
      invitedUserName: "Invited User",
    },
    getTestSession({ user: inviter })
  )
).rejects.toThrow("User is not logged into any organization")
```

```tsx
const orgId = 123;

const result = await inviteUser(
  {
    asRole: "USER",
    invitedUserEmail: "inviteduseremail@test.com",
    invitedUserName: "Invited User",
  },
  getTestSession({ user: inviter, orgId })
)
// should succeed
```

**Reply by Talor-A** [reply]
actually, I'm realizing this may not answer the question as I don't have test assertions against `resolver.authorize()`.

**Reply by flybayer** [reply]
@Talor-A what assertions are you trying to test? Maybe share an entire test case that you want to work but isn't.

**Reply by Talor-A** [reply]
I'd want to test that the session passed to `inviteUser` would cause the mutation to throw without `role:ADMIN` set. Is there a way to do this?

**Reply by flybayer** [reply]
@Talor-A yes, for sure. Can you share the code of your best attempt at this? Then we can modify for you.

**Comment by flybayer** [comment]
Fyi, I opened an issue on making this easier to do: https://github.com/blitz-js/blitz/issues/2654

**Comment by vincentrolfs** [comment]
Thanks a lot, my original issue is solved. I described the solution here: https://github.com/blitz-js/blitz/issues/2654#issuecomment-908475727
